### ISSUE TITLE
fail of allocation

### ISSUE THREAD

**Issue by kirill-pinigin** [tiny-dnn/tiny-dnn#709]
std::vector<label_t> prepareLabels(const int num, const label_t label)
{
std::vector<label_t> result;
for (int i = 0; i <num; ++i)
result.push_back(label);
return result;
}

std::vector<vec_t> ConvertImage(const std::string & filename)
{
cv::Mat bigImage = cv::imread(filename, cv::IMREAD_GRAYSCALE);
const int batchSize = 32;
cv::Size smallSize(batchSize, batchSize);
std::vectorcv::Mat smallImages;
if (bigImage.rows < batchSize || bigImage.cols < batchSize)
throw std::invalid_argument("small size of image \n");
for (int y = 0; y < bigImage.rows; y += smallSize.height)
{
for (int x = 0; x < bigImage.cols; x += smallSize.width)
{
cv::Rect rect = cv::Rect(x, y, smallSize.width, smallSize.height);
smallImages.push_back(cv::Mat(bigImage, rect));
}
}
double scale = 1.0;
std::vector<vec_t> data;
vec_t d;
for (auto & item : smallImages) {
vec_t d;
std::transform(item.datastart, item.dataend, std::back_inserter(d),
[=](uint8_t c) { return c * scale; });
data.push_back(d);
}
return data;
}

int main(int argc, char** argv)
{
#define O true
#define X false
network net;
layers::input in(shape3d(32, 32, 1));
convolutional_layer conv(32, 32, 7, 7, 1, 400 * 1, padding::valid, true, 1, 1);/* 32x32 in, 7x7 kernel, 1-400 fmaps conv /
average_pooling_layer av_p(26, 26, 400 * 1, 2, 2, 2, 2, padding::valid);/ 26x26 in, 400 fmaps, 2x2 subsampling /
max_pooling_layer max_p(26, 26, 400 * 1, 2, 2, 2, 2, padding::valid);/ 26x26 in, 400 fmaps, 2x2 subsampling */
layers::fc fc1(67600 * 1, 1024), fc2(67600 * 1, 1024);
layers::concat concat(2, 1024);
layers::dense dense1(2048, 2048), dense2(2048, 10);
in << conv;
conv << av_p;
conv << max_p;
av_p << fc1;
max_p << fc2;
(fc1, fc2) << concat;
concat << dense1;
dense1 << dense2;

construct_graph(net, { &in }, { &dense2 });

adagrad optimizer;

auto samples = ConvertImage("E:/testimage2.bmp");
auto targets = prepareLabels(samples.size(), 80);
net.train<mse>(optimizer, samples, targets, 1, 20, nop, nop);
tensor_t input(32, vec_t(32, 128));
auto output = net.predict(input);
    return 0
}

I catch exception 
pointer allocate(size_type size, const void * = nullptr) {
    void *p = aligned_alloc(alignment, sizeof(T) * size);
    if (!p && size > 0) throw nn_error("failed to allocate");
    return static_cast<pointer>(p);
  }

**Comment by kirill-pinigin** [comment]
also  I want to notice  this problem  is actual only in graph model of neural net!! When I use sequential model  I don't have this problem

network<sequential> net;
	//layers::input in(shape3d(32, 32, DIM));
	convolutional_layer conv(32, 32, 7, 7, DIM, 400 * DIM, padding::valid, true, 1, 1);
	average_pooling_layer av_p(26, 26, 400 * DIM, 2, 2, 2, 2, padding::valid);
	max_pooling_layer max_p(26, 26, 400 * DIM, 2, 2, 2, 2, padding::valid);
	layers::fc fc1(67600 * DIM, 1024), fc2(67600 * DIM, 1024);
	layers::concat concat(2, 1024);
	layers::dense dense1(2048, 2048), dense2(1024, 10);
	net << conv << av_p << fc1<<activation::tanh()<< dense2 << activation::softmax();
std::vector<vec_t> samples;
samples.emplace_back(vec_t(32 * 32, 128));
samples.emplace_back(vec_t(32 * 32, 256));
//std::vector<vec_t> targets(samples.size(), vec_t(1, 100));
std::vector<label_t> targets{ 7,8 };
net.train<cross_entropy> (optimizer, samples, targets,1,1);
vec_t input (32*32, 129);
auto output = net.predict(input);
return 0;

}

**Comment by Randl** [comment]
The following code runs ok at my PC both with and without AVX:
```cpp
#include <iostream>
#include "tiny_dnn/tiny_dnn.h"

using namespace tiny_dnn;
using namespace tiny_dnn::activation;
using namespace std;

int main(int argc, char **argv) {
  network<graph> net;
  layers::input in(shape3d(32, 32, 1));
  convolutional_layer
      conv(32, 32, 7, 7, 1, 400 * 1, padding::valid, true, 1, 1);
  average_pooling_layer av_p(26, 26, 400 * 1, 2, 2, 2, 2, padding::valid);
  max_pooling_layer max_p(26, 26, 400 * 1, 2, 2, 2, 2, padding::valid);
  layers::fc fc1(67600 * 1, 1024), fc2(67600 * 1, 1024);
  layers::concat concat(2, 1024);
  layers::dense dense1(2048, 2048), dense2(2048, 10);
  in << conv;
  conv << av_p;
  conv << max_p;
  av_p << fc1;
  max_p << fc2;
  (fc1, fc2) << concat;
  concat << dense1;
  dense1 << dense2;

  construct_graph(net, {&in}, {&dense2});

  adagrad optimizer;

  std::vector<vec_t> samples;
  samples.emplace_back(vec_t(32 * 32, 128));
  samples.emplace_back(vec_t(32 * 32, 256));
  //std::vector<vec_t> targets(samples.size(), vec_t(1, 100));
  std::vector<label_t> targets{7, 8};
  net.train<cross_entropy>(optimizer, samples, targets, 1, 1);
  vec_t input(32 * 32, 129);
  auto output = net.predict(input);
  for (auto x: output) {
    std::cout << x << " ";
  }
  return 0;

}
```

**Comment by kirill-pinigin** [comment]
Randl, what is your operating system? How are you think ? Where can I find this exception? if you don't confirm this case?

**Comment by kirill-pinigin** [comment]
![image](https://cloud.githubusercontent.com/assets/27665392/26041399/b6258bfc-3956-11e7-896a-c407b068c6c6.png)

**Comment by kirill-pinigin** [comment]
![image](https://cloud.githubusercontent.com/assets/27665392/26041436/fb1a1f34-3956-11e7-83e6-869f20300629.png)

**Comment by kirill-pinigin** [comment]
![image](https://cloud.githubusercontent.com/assets/27665392/26041444/0d9f230c-3957-11e7-8c29-7758e2962cbd.png)

**Comment by kirill-pinigin** [comment]
Do you use OpenCL ?What  the options of cmake do you use?

**Comment by Randl** [comment]
I don't have visual studio, so I can't help much. It seems to be platform dependent. 
cc @beru

**Comment by kirill-pinigin** [comment]
Randl, yes you are right !  I check this example in Ubuntu 16.04.02  And it is works.

**Comment by beru** [comment]
@Randl I don't always have Windows PC, so can't be of helf right now.
